### DISCUSSION TITLE
使用jsdom解决第三方插件库不支持SSR所带来的新问题！

### DISCUSSION THREAD

**Discussion by xsdream** [vuepress/core#931] [Q&A] [answered]
在vuepress.config.js的头部我增加了以下代码，第三方插件库不再报找不到window下面变量的错误了。
```js
const jsdom = require("jsdom");
const {JSDOM} = jsdom;
const dom = new JSDOM('<!DOCTYPE html><html><body></body></html>')
if (typeof window === 'undefined') {
    global.window = dom.window;
    global.document = dom.window.document;
    global.navigator = dom.window.navigator;
    global.location = dom.window.location;
}
```
但是他带来了新的问题：
![image](https://user-images.githubusercontent.com/31870384/172341212-5b3fa2c9-dc39-4f5d-aaf7-ba2affb45748.png)

**Comment by xsdream** [comment]
d-comp\node_modules\sass\sass.dart.js 
我检查了 sass.dart.js这个文件，发现里面做了对node和web环境的判断，是不是因为我强行在node环境下给global.window赋值导致了这个问题，在vuepress中可否解决？

**Comment by xsdream** [comment]
@Mister-Hope 大佬可以帮忙看一下吗，之前我在issues中提交过类似的问题

**Comment by Mister-Hope** [comment, marked as answer]
不是，我都没搞清楚你在干什么？

- 你用 JSDOM 模拟一个 document 全局对象有**任何意义**么？没有，因为这并不是客户端的 DOM。
- 在执行 SSR 的时候，这些库有 **实际上渲染出什么东西来么**？没有，因为它们调用的是一个无关的虚拟 DOM。
- 那你在 SSR 期间执行调用这些库的操作，有**任何意义**么？没有。

对于所有涉及操作 DOM 的库，你都应该在 Client-Only 周期或仅在客户端侧挂载的组件中动态载入。换而言之，只有在用户浏览器上，VuePressApp 挂载之后，你才应该**导入并解析这些库，之后调用对应方法更新DOM**。

听懂了么？

**Comment by xsdream** [comment]
首先表示感谢，今天终于把问题解决了。
1. 使用JSDOM 是因为在stackoverflow看到一个帖子的恢复，说是用JSDOM能解决第三方包在SSR下访问client属性的问题，也没有做太多的了解就病急乱投医了
2. 使用await import的方式我在看到官网【使用不支持 SSR 的功能】的时候就实验过了，但是渲染页面的时候报未找到第三方插件方法的错误，我并没有想到用<ClientOnly>,因为当时我是页面用的是Markdown和vue SFC结合的方式，现在回去看官网，已经写的很详细了
![image](https://user-images.githubusercontent.com/31870384/172519025-89876453-939e-41ae-800e-e4af2aa0ae60.png)
3. 反思自己理解问题和解决问题的能力确实有待加强
4. 最后向你表示感谢，很耐心的在issues和discussion回复我
@Mister-Hope
